lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsService.cs

```diff
@@ -54,24 +75,34 @@ public void OnNext(IContextMessage contextMessage)
             var counters = new EvaluatorMetrics(msgReceived).GetMetricsCounters();
             Logger.Log(Level.Info, "Received {0} counters with context message: {1}.", counters.GetCounters().Count(), msgReceived);
 
-            foreach (var counter in counters.GetCounters())
+            _countersData.Update(counters);
+
+            if (_countersData.TriggerSink(_counterSinkThreshold))
             {
-                ICounter c;
-                if (_counters.TryGetValue(counter.Name, out c))
+                Sink(_countersData.GetCounterData());
+                _countersData.Reset();
+            }
+        }
+
+        /// <summary>
+        /// Call each Sink to sink the data in the counters
+        /// </summary>
+        private void Sink(ISet<KeyValuePair<string, string>> set)
+        {
+            foreach (var s in _metricsSinks)
+            {
+                try
                 {
-                    //// TODO: [REEF-1748] The following cases need to be considered in determine how to update the counter:
-                    //// if evaluator contains the aggregated values, the value will override existing value
-                    //// if evaluator only keep delta, the value should be added at here. But the value in the evaluator should be reset after message is sent
-                    //// For the counters from multiple evaluators with the same counter name, the value should be aggregated here
-                    //// We also need to consider failure cases.  
-                    _counters[counter.Name] = counter;
+                    Task.Run(() => s.Sink(set));
                 }
-                else
+                catch (Exception e)
                 {
-                    _counters.Add(counter.Name, counter);
+                    Logger.Log(Level.Error, "Exception happens during the sink for Sink {0} with Exception: {1}.", s.GetType().AssemblyQualifiedName, e);
+                }
+                finally
+                {
+                    s.Dispose();
```

Comment:
not possible - tang guards it

---
In reply to: [111804538](https://github.com/apache/reef/pull/1262#discussion_r111804538) [](ancestors = 111804538)